### PR TITLE
fixes #4201 - update operating system by label

### PR DESCRIPTION
**Pull request by isratrade**
operatingsystem_names uses `to_label` which returns description if it exists or fullname (name, major, minor)

### CODE REVIEW

File: app/models/operatingsystem.rb [hunk 1]
```diff
@@ -131,8 +132,9 @@ def fullname
     to_s
   end
 
-  def self.find_by_fullname(fullname)
-    a = fullname.split(" ")
+  def self.find_by_to_label(str)
+    return @os if (@os = Operatingsystem.find_by_description(str))
```
Comment (domcleal) [inline app/models/operatingsystem.rb]: nitpick, no need for an instance variable here?

Comment (domcleal) [inline app/models/operatingsystem.rb]: s/Operatingsystem/self/ ?

**Comment by domcleal** [comment]
I've tried doing a `GET /api/v2/operatingsystems/foo` (where an OS has its description set to `foo`) but the OS isn't found.  SQL logs show it's trying to search by name and not finding it.

Do we need to define resource_identifying_attributes with `to_label` instead for it to work?

Also I think routes might need changing to accept names containing full stops, e.g. `CentOS 6.5` which comes through to the controller as `CentOS 6`.

File: app/models/operatingsystem.rb [hunk 2]
```diff
@@ -119,6 +120,11 @@ def to_label
     description.blank? ? to_s : description
   end
 
+  # to_label setter updates description and does not try to parse and update major, minor attributes
+  def to_label=(str)
+    self.description = str
```
Comment (isratrade) [inline app/models/operatingsystem.rb:126]: @domcleal, added this setter method and test and re-committed.

**Comment by isratrade** [comment]
@domcleal, recommited
- edited routes to allow id with spaces or dots.  api/v2/operatingsystems/RHEL%206.3
- added `to_label` to resource_identifying_attributes
- added some tests

File: app/models/operatingsystem.rb [hunk 3]
```diff
@@ -116,23 +117,31 @@ def interpolate_medium_vars path, arch, os
...
-    to_s
+  def to_s
+    fullname
   end
 
-  def self.find_by_fullname(fullname)
-    a = fullname.split(" ")
+  def self.find_by_to_label(str)
+    os = self.find_by_description(str)
+    return os if os
```
Comment (isratrade) [inline app/models/operatingsystem.rb:143]: not sure why this one-liner didn't work

``` ruby
    return os if (os = self.find_by_description(str))
```

I get the error

```
NameError: undefined local variable or method `os' for #<Class:0x0000000afb4a78>
```

so I split it into two lines and it works ????

**Comment by domcleal** [comment]
@isratrade please rebase

**Comment by isratrade** [comment]
rebased and re-commited

**Comment by domcleal** [comment]
Merged as 42c9541 for Foreman 1.5.0, thanks @isratrade!

**Pull request closed without merging**